### PR TITLE
Fix race conditions during exiting.

### PR DESCRIPTION
**Pull request by jianjunz**
A side effect of this change is memory usage increase because it moves ended channel to a vector instead of dispose them immediately.

This change also add an reference to `pc_factory_ `when it is created. It causes memory leak but fixes hang issue when application exits. AtExitManager might be needed to eventually fix this issue.

### CODE REVIEW

File: talk/owt/sdk/include/cpp/owt/p2p/p2pclient.h [hunk 1]
```diff
@@ -222,6 +222,7 @@ class P2PClient final
   std::shared_ptr<P2PSignalingChannelInterface> signaling_channel_;
   std::unordered_map<std::string, std::shared_ptr<P2PPeerConnectionChannel>>
       pc_channels_;
+  std::vector<std::shared_ptr<P2PPeerConnectionChannel>> removed_pc_channels_;
```
Comment (HuaChunbo) [inline talk/owt/sdk/include/cpp/owt/p2p/p2pclient.h]: What is removed_pc_channels_ variable for?

Comment (jianjunz) [reply talk/owt/sdk/include/cpp/owt/p2p/p2pclient.h]: `removed_pc_channels` includes closed channels. Disposing `PeerConnectionChannel` immediately when ICE connection state changes to closed causes deadlock.

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 2]
```diff
@@ -347,6 +347,10 @@ void P2PClient::OnMessageReceived(const std::string& remote_id,
                          &P2PClientObserver::OnMessageReceived,
                          remote_id, message);
 }
+void P2PClient::OnStopped(const std::string& remote_id) {
+  removed_pc_channels_.push_back(pc_channels_[remote_id]);
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc]: Please associate the vector push_back/erase with lock.

Comment (jianjunz) [reply talk/owt/sdk/p2p/p2pclient.cc]: Done.

**Comment by jianjunz** [comment]
Fixes #190.

**Pull request merged**